[Data] Cap op concurrency with exponential ramp-up

raulchen opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Today when a new dataset is launched, the StreamingExecutor will allocate all resources to the first operator. 
Consider a simple case `read -> map` where the read and map are not fused, the scheduler will submit as many read tasks as possible up front. Assuming the read tasks will output many blocks, the blocks will pile up because we don’t have resources to submit map tasks to consume the data.

This PR tries to mitigate this issue by cap the concurrency of each op with an initial value and ramp up the cap exponentially as the execution goes on (see `ConcurrencyCapBackpressurePolicy` docstring for details). 

This PR also introduces a `BackpressurePolicy` interface, making the backpressure policies configurable and pluggable. Later we should migrate existing backpressure mechanisms to this new interface.

Known limitations:
- If the config is not properly set, perf may regress for some workloads. Thus we disable this feature by default in 2.8, and will enable it in 2.9.
- This feature only caps the initial concurrency. Once the cap has ramped up, data can still pile up. This issue will be resolved by a different backpressure policy that will profile the runtime metrics (e.g, object store increase, RAM usage) of each op. 
- It doesn't backpressure tasks that output many blocks. This issue will be solved by streaming generator backpressure instead. 
 

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

c21 commented on python/ray/data/_internal/execution/backpressure_policy.py:
```diff
@@ -0,0 +1,100 @@
+
+    @abstractmethod
+    def can_run(self, op: "PhysicalOperator") -> bool:
+        """Called when StreamingExecutor is about to select an operator to run.
```
Shall we be more specific for which function will be calling it, such as 

```
Called when StreamingExecutor selects an operator to run in `streaming_executor_state.select_operator_to_run()`.
Returns True if the operator can run, False otherwise.
```

c21 commented on python/ray/data/_internal/execution/backpressure_policy.py:
```diff
@@ -0,0 +1,100 @@
+    It will be set to an intial value, and will ramp up exponentially.
+
+    The concrete stategy is as follows:
+    - Each op is assigned an initial concurrency cap.
```
nit: `op` -> `PhysicalOperator` and other places.

c21 commented on python/ray/data/_internal/execution/backpressure_policy.py:
```diff
@@ -0,0 +1,100 @@
+
+    # Environment variable to configure this policy.
+    # The format is: "<init_cap>,<cap_multiply_threshold>,<cap_multiplier>"
+    CONFIG_ENV_VAR = "RAY_DATA_CONCURRENCY_CAP_CONFIG"
```
this is not blocking comment, but better to have separate environment variables.

raulchen replied:
This is to facilitate internal tests in 2.9. When we officially release this feature, we will probably simplify the configs. So I'll keep it for now.

c21 commented on python/ray/data/_internal/execution/backpressure_policy.py:
```diff
@@ -0,0 +1,100 @@
+    # The format is: "<init_cap>,<cap_multiply_threshold>,<cap_multiplier>"
+    CONFIG_ENV_VAR = "RAY_DATA_CONCURRENCY_CAP_CONFIG"
+
+    # The intial concurrency cap for each operator.
+    INIT_CAP = 4
+    # When the number of finished tasks reaches this threshold, the concurrency cap
+    # will be multiplied by the multiplier.
+    CAP_MULTIPLY_THRESHOLD = 0.5
+    # The multiplier to multiply the concurrency cap by.
+    CAP_MULTIPLIER = 2.0
```
another way is to define these constant on the file level, so we don't need to parse environment variables. It's also maybe easier for advanced users to test out.

raulchen replied:
The issue with constants is that if the executor doesn't run on the driver (e.g., on the SplitCoordinator actor), it's hard to change the configs. I've seen this issue for other configs that depend on constants. Do you know a good solution to bypass this issue?

stephanie-wang replied:
I think you'll need to put them into the DataContext.

raulchen replied:
@stephanie-wang @c21 If we put them into DataContext, I'd like to to save them in a dict and add a key-value interface in DataContext. 
The reason is because this is a plugin and DataContext shouldn't need to know about the plugin configs. 
What do you think?

The API will be something like:
```python
data_context.set("concucrrency_cap_backpressure_policy.init_cap", 4)
```

stephanie-wang replied:
That sounds good to me for now while it's still experimental. Actually, could you prepend the name with "experimental" or something like that? Makes deprecation a bit smoother.

c21 commented on python/ray/data/_internal/execution/backpressure_policy.py:
```diff
@@ -0,0 +1,100 @@
+
+
+# TODO(hchen): Enable ConcurrencyCapBackpressurePolicy by default.
+DEFAULT_BACKPRESSURE_PLOCIES = []
```
So do we foresee in the near future, we will have multiple policies enabled at same time?

raulchen replied:
I do plan to migrate existing backpressure code to this interface. 
Another reason why I wanted to introduce this interface is that it will allow us experimenting new back-pressure policies without touching the code base. E.g., one idea is to take the real runtime metrics into consideration for backpressure.

stephanie-wang replied:
Sounds good, I don't feel strongly about it, just think that it may be too early to abstract (we don't know yet if this is the right interface).

raulchen replied:
yes, the interface is likely to change as we implement other policies. it's an internal interface, so that should be fine.

c21 commented on python/ray/data/_internal/execution/backpressure_policy.py:
```diff
@@ -0,0 +1,100 @@
+
+
+# TODO(hchen): Enable ConcurrencyCapBackpressurePolicy by default.
+DEFAULT_BACKPRESSURE_PLOCIES = []
```
`PLOCIES` -> `POLICIES`

c21 reviewed (commented):
LG overall.

stephanie-wang commented on python/ray/data/_internal/execution/streaming_executor_state.py:
```diff
@@ -377,6 +378,7 @@ def select_operator_to_run(
     topology: Topology,
     cur_usage: TopologyResourceUsage,
     limits: ExecutionResources,
+    back_pressure_policies: List[BackpressurePolicy],
```
```suggestion
    backpressure_policies: List[BackpressurePolicy],
```

stephanie-wang reviewed (commented):
I'm fine merging if we turn it off by default, but I don't think this policy is going to be usable. The performance regression is scary, and the configuration parameters here are going to be pretty confusing for the average user, I think. Also, off the top of my head, there are likely many cases that won't work or will regress (e.g., concurrency=4 cap may or may not work depending on num_cpus/task and num_cpus total).

Offline, let's try to come up with a policy that minimizes regressions and has as few configuration parameters as possible, ideally none. I believe we can come up with something reasonable by considering the total number of operators when assigning the initial caps. For example, if we have N operators to run, we give each 1/N cores initially then adjust based on which operators are ready.

Also, I think this PR needs some doc changes to instruct on how to turn on the feature and configure it.

stephanie-wang commented on python/ray/data/tests/test_backpressure_policies.py:
```diff
@@ -0,0 +1,81 @@
+from ray.data._internal.execution.streaming_executor_state import Topology
+
+
+class TestConcurrentcyCapBackpressurePolicy(unittest.TestCase):
```
```suggestion
class TestConcurrencyCapBackpressurePolicy(unittest.TestCase):
```

raulchen commented:
@stephanie-wang You are right. The current default config isn't optimal. I also thought of making the default config smarter by taking into consideration the number of ops and their resource requirements. The plan is to implement the basic framework and turn this off in 2.8. In 2.9, we'll need to do more experiments to figure out the best configuration and officially release this feature.

raulchen commented:
@c21 @stephanie-wang thanks for your comments. They are either addressed (resolved threads) or replied (non-resolved threads). Please take a look again.

stephanie-wang reviewed (approved):
Let's merge after making it configurable through DataContext instead of the env variables.

raulchen commented:
Moved the config to DataContext and added the `set_plugin_config/get_plugin_config` API. Reason was commented [here](https://github.com/ray-project/ray/pull/40275#discussion_r1362544555).

c21 commented on python/ray/data/context.py:
```diff
@@ -219,6 +219,7 @@ def __init__(
         self.enable_get_object_locations_for_metrics = (
             enable_get_object_locations_for_metrics
         )
+        self._plugin_configs: Dict[str, Any] = {}
```
can we rename it as `_backpressure_plugin_configs`? In the future, we may introduce other plugin components.

c21 commented on python/ray/data/context.py:
```diff
@@ -283,6 +284,15 @@ def _set_current(context: "DataContext") -> None:
         global _default_context
         _default_context = context
 
+    def get_plugin_config(self, key: str, default: Any = None) -> Any:
+        return self._plugin_configs.get(key, default)
+
+    def set_plugin_config(self, key: str, value: Any) -> None:
+        self._plugin_configs[key] = value
+
+    def remove_plugin_config(self, key: str) -> None:
+        self._plugin_configs.pop(key, None)
```
let's declare these methods with `_` prefix, so they remain private and easy to change later.

raulchen replied:
Discussed offline, we'll make this API reusable for other components.

c21 reviewed (approved):
LG w/ two minor comments.
